### ISSUE TITLE
[Bug] ProtonMail mobile app stalls on everything after large email & contacts import

### ISSUE THREAD

**Issue by saizai** [ProtonMail/proton-mail-android#195]
**Describe the bug**

After importing a large amount of email, contacts, etc (~200k email threads, ~20GB, ~300 labels, ~10k contacts) from Gmail using Proton Mail Easy Switch, the Android app stalls on startup, sync, direct invocation of message composition via the home screen widget, etc. These were all working fine before the import, and broke after, so I'm fairly sure that's the cause.

The website version, on Android Chrome, does not have this problem. I have not tested other clients.

**To Reproduce**
1. Use the Android app
2. Go to https://account.proton.me/u/0/mail/easy-switch and import an account with a large amount of emails, contacts, labels, etc. 
3. Try to use the Android app and have everything freeze
4. Notice that the sync notification is also frozen, and there are no "new mail received" notifications 
5. Notice that in the web app, opening the sidebar is slow (and hard to scroll) due to large number of flattened labels. (See note on Easy Switch in suggested fixes below.)

**Expected behavior**

UX: Should work as fast as before import. 

Code:

My guess is that the problem is from any combination of a few things:
* trying to store a local index of the whole mailbox
* trying to keep all gmail-imported labels in RAM
* trying to keep all contacts in RAM
* … etc

I suggest, in increasing complexity:
* Lazy load things wherever possible.
* Save user input (eg draft emails) before doing anything that might be a stallingly-large select.
* Separately store the N most recently used items in each category, require intentional action by user to load more.
* Store recent mail, labels, contacts, etc in some sort of RAM-optimised manner to minimise both cache misses and unneeded cache size (eg a prefix tree if the rest can fit on SD card).
* During Easy Switch, and in labels/folders editors, allow user to easily map imported labels/folders to existing ones. (Currently, it actually _prohibits_ using an imported label that is identical to an existing label, and it prohibits nested/labels/like/this, transforming them to not-nested-labels-like-this, which makes for way more items at the root level. Which means I'll have to run sieves to re-label everything,)
* Auto sort labels & folders by alpha unless told not to. Make "sort by" an option, default "alpha", other options: "manual" (i.e. current behaviour), "date added" (i.e. label made), "date updated" (e.g. most recent email with this label received). In UX, encourage use of nested (i.e. /-delimited) labels & folders.
* Explicitly treat RAM as a cache for SD card as a cache for web request, with RAM on heavy limits against ballooning.
* Meta-tag contacts, labels, emails, etc as directly user-made (e.g. typed by hand, interacted with, etc) vs bulk. Don't autoload the bulk ones where avoidable (exception: recent email).
* Use label/folder tree structure for RAM saving — don't keep unexpanded nodes in RAM.
* Run a performance check on phone to find out how many labels, contacts, emails, whatever causes it to take an excessive performance hit (just generate fake ones locally for test, co-process to check wall time for lookup, insert, etc), and use the results to tune limits on how many to store in RAM / SD card.
* Have a continuous reaper co-process that monitors whether the main process is taking too long, going into swap, causing device to overheat, or the like — especially for core operations like sync, read recent mail, and compose mail — and have it early abort _that operation_ without affecting anything else if possible (e.g. if it's a list operation, return a limited # of results, with UI saying so); if not possible, kill and auto-file bug report stating the performance, operation type, and progress at which it stalled (should be easy to anonymise). Make sure the kill doesn't harm user input like message composition; if anything it should to the opposite (eg if composing a message triggers an over-limit contacts lookup, early terminate the lookup and let the user keep writing, and save the draft).

**Smartphone (please complete the following information):**
 - Device: Samsung Galaxy S9
 - OS: Android 10

**Comment by captainnapalm** [comment]
Yes, I've experienced the exact same issue. Same steps to reproduce. This only happened after doing a large import from Gmail for mail and contacts.  The stuck notification "Retrieving notifications" causes massive battery drain and battery/device temperature skyrockets whenever I open Proton Mail on android. I have to open the app, let Proton refresh and check emails, then force close the app to avoid it burning a hole in my pocket and depleting my battery.

Device: Fairphone 4 5G
OS: Android 11

**Comment by NWinocur** [comment]
Have experienced same on a Pixel 6 running Android 13.  App has felt nigh-unusably slow after import via "easy switch."  I can be alongside a colleague spending literally the entire duration of a feature-length film trying to search for, read, compose a reply to, and sending one single email while they take care of their entire day's work from their own mobile device running an altogether different email client.

**Comment by captainnapalm** [comment]
The app doesn't seem to be able to keep up with updates on the web after Easy Switch. I've moved almost everything in my inbox to Archive, and the mobile app still shows most emails as being in the Inbox and takes minutes to refresh and then still not do anything.

**Comment by captainnapalm** [comment]
Not sure if something changed server side, but my app has caught up to changes on the web, doesn't seem to drain battery anymore, and is more responsive. Don't think there's been a recent app update to correlate this to.

**Comment by brownan** [comment]
New subscriber to protonmail here, and I've noticed this too after a big gmail import. Messages archived from the web interface don't seem to archive on the android app, the app is particularly slow to refresh various mailbox views, battery is draining unreasonably fast, and there's an android notification from the app (in the android pull-down menu) that says "Proton Mail Retrieving notifications" which can't be dismissed.

I'm assuming the android app is doing some really hefty sync work in the background, and for the duration it causes those symptoms. If so, I hope it'll finish soon.

**Comment by brownan** [comment]
Logging out of the app, clearing its cache, then logging back in seems to have fixed the issue for me.

**Comment by mvhirsch** [comment]
> After importing a large amount of email, contacts, etc (~200k email threads, ~20GB, ~300 labels, ~10k contacts) from Gmail using Proton Mail Easy Switch, the Android app stalls on startup, sync, direct invocation of message composition via the home screen widget, etc. These were all working fine before the import, and broke after, so I'm fairly sure that's the cause.

Same thing on my device after importing from Gmail using Easy Switch. I imported about 91k Emails (1.43 GB). Since I practice zero-inbox I then needed to archive all imported Emails (which took quite a time). But the app still showed emails in my inbox. Switching to a different view (e.g. "marked" or "all") took several minutes(!). Sending an email took 2-3 minutes (time between pressing send-button and notification "sent successfully").

> Logging out of the app, clearing its cache, then logging back in seems to have fixed the issue for me.

This resolved my issues! 

So I can confirm, there is something in the sync-process, which causes lots of delays. After a fresh login, it works flawlessly.

Device: Nokia 8.3, Android 12
App-Version: 3.0.13 (934)
